pkg/sessions/redis/redis_store.go

```diff
@@ -65,6 +67,69 @@ func (store *SessionStore) Clear(ctx context.Context, key string) error {
 	return nil
 }
 
+// ClearAll clears all saved sessions' information for a given user
+// from redis, and then clears the session
+func (store *SessionStore) ClearAll(ctx context.Context, s *sessions.SessionState, exp time.Duration, password string, user string) error {
+	if password != os.Getenv("OAUTH2_PROXY_ADMIN_PASS") {
+		return errors.New("access denied")
+	}
+	keyName := fmt.Sprintf("sessionlist-%s", user)
+	err := s.ObtainLock(ctx, exp)
+	if err != nil {
+		return fmt.Errorf("error occurred while trying to obtain lock: %v", err)
+	}
+	defer func() {
+		if s == nil {
+			return
+		}
+		if err := s.ReleaseLock(ctx); err != nil {
+			logger.Errorf("unable to release lock: %v", err)
+		}
+	}()
+	value, err := store.Load(ctx, keyName)
+	if err != nil {
+		return err
+	}
+	for _, sessionKey := range strings.Split(string(value), ":") {
+		err = store.Client.Del(ctx, sessionKey)
+		if err != nil {
+			return fmt.Errorf("error clearing the session %v from redis: %v", sessionKey, err)
+		}
+	}
+	err = store.Client.Del(ctx, keyName)
+	if err != nil {
+		return fmt.Errorf("error clearing the session %v from redis: %v", keyName, err)
+	}
+	return nil
+}
+
+func (store *SessionStore) SaveUserSession(ctx context.Context, s *sessions.SessionState, value string, exp time.Duration) error {
```

Comment:
We also need a similar function called from the persistance manager Clear function, to remove a session key from this document when the session is removed.

Reply:
@joel-bluedata We are currently on it. Currently clear function does not have access to SessionState object, we need SessionState object to find the user which signs out using SessionState.User.

Tried using 

session := middlewareapi.GetRequestScope(req).Session to get sessionState details but it returns nil in ClearSession function we tried to create.

Currently clearAll endpoint will get the session list key that we are creating and delete all the session keys within it **even if there are stale entries.** These stale entries can be due to sign out or any out of band removal of the session

Finally session list key for the user will be deleted.